src/utils.jl

```diff
@@ -452,16 +452,45 @@ end
 
 
 function flatten1(ex)
+  isexpr(ex, :try) && return flatten_tryblock(ex)
   isexpr(ex, :block) || return ex
   #ex′ = :(;)
   ex′ = Expr(:block)
   for x in ex.args
     isexpr(x, :block) ? append!(ex′.args, x.args) : push!(ex′.args, x)
   end
-  # Don't use `unblock` to preserve line nos
+  # Don't use `unblock` to preserve line nos.
   return length(ex′.args) == 1 ? ex′.args[1] : ex′
 end
 
+function flatten_tryblock(ex) # undo the damage potentially done by `flatten1` - the main problem is that `try` expects expressions as arguments, but `quote false end` might get flattened to `false`, leading to ambiguity
+    if length(ex.args) == 3
+        # try catch end # 3 args, 2nd is a symbol or false, 3rd is expr
+        !isexpr(ex.args[3]) && (ex.args[3] = quote $(ex.args[3]) end)
+    elseif length(ex.args) == 4
+        # try finally end # 4 args, 2nd==3rd==false, 4th is expr
+        # try catch finally end # 4 args, 2nd is a symbol or false, 3rd and 4th are expr
```

Comment:
For review purposes, **before this PR**, the following was being processed incorrectly (the error is not caught):

```
quote; try; error(); catch; false; finally; println(123); end; end |> striplines |> flatten |> eval
```

Now `flatten` raises an error and refuses to proceed.

However, the following used to work, but now it raises an error (with a workaround suggested in the error message):

```
quote; try; 1+1; finally; println(123); end; end |> striplines |> flatten |> eval
```

I do not see how to avoid the silent bug in the first snippet without refusing to process the second snippet as well -- well, except for a full rewrite of `flatten` that does not use `postwalk`.

Reply:
I think making sure people do not run into silent bugs is a good enough reason to sacrifice a bit of related accidentally working functionality.

Reply:
It is also of note that this is highly improbable to be triggered: someone needs to explicitly use `striplines` otherwise the line number nodes prevent this ambiguity from happening.

Reply:
>  except for a full rewrite of flatten that does not use postwalk.

I think it'd be worth giving this a try, in a separate PR. Sorry that I said I wouldn't like that approach; I just didn't see a satisfying solution anywhere. Mike Innes' original code is beautifully simple and straight-forward, and I'd like to keep that as much as possible. 

From my perspective, Julia's AST design is at fault here, but there's no way we'll get that changed...

In ay case, I would start by either:

- copy-pasting post-walk and adding from there.
- Trying with `prewalk`. Maybe there we can do something to the third element of `try` to make sure that it doesn't collapse to `false`. But I don't think it'll be that satisfying either, so we might as well do the correct solution with a custom tree traversal.

Thank you for tackling this!

Reply:
Something like

```julia
flatten(x) = x
function flatten(x::Expr)
  if isexpr(x, :try)
      Expr(:try, flatten(x.args[1]), flatten(x.args[2]), Expr(:block, map(flatten, x.args[3].args)...))
   else
     flatten(Expr(x.head, map(flatten, x.args)...))
end
```

(just a sketch, this is full of mistakes for sure)

In retrospect, that's quite simple and we get to solve all problems perfectly. Sorry for not having suggested it sooner.

Reply:
I am missing something. I was under the impression that only the following options exist for try blocks:

- 3 argument blocks where the 3rd argument is a block, corresponding to `try catch end` 
- 4 argument blocks where the 3rd argument is a block, corresponding to `try catch finally end`
- 4 argument blocks where the 3rd argument is false, corresponding to `try finally end`
- 5 argument blocks where the 4th argument is a block, corresponding to `try catch else finally end`
- 5 argument blocks where the 4th argument is false, corresponding to `try catch else end`

In your example above it seems like you only permit 3 argument blocks?

Or maybe you did not mean to focus on specifically what the special case for try should look like, rather you are telling me to use a "prewalk" style recursion instead of the current "postwalk" style recursion? I think I agree with that, but I will not have the time to attempt to contribute this for a couple of weeks.

Reply:
> Or maybe you did not mean to focus on specifically what the special case for try should look like, rather you are telling me to use a "prewalk" style recursion instead of the current "postwalk" style recursion?

Yes, that. Of course we need to handle all the cases you mentioned. Sorry that wasn't clear.

It's not a prewalk, in the sense that it doesn't apply the transformation again on the outcome. It's like postwalk, where we simply do not directly apply the transformation to the third argument of the try, but deal with it appropriately. 

> I think I agree with that, but I will not have the time to attempt to contribute this for a couple of weeks.

👍 Thank you very much for the PR and issue! It helps a lot. I don't think that Julia 1.10 will be ready for a long time, so I don't think it's very pressing. Could even be fixed after release, arguably.